Introduce Policies-Service (#5716)
owncloud/ocis: services/policies/README.md

> # Policies Service

Reviewer: Please add something. If you just copy the description from the changelog we will already have a good starting point.

# Policies Service

> The Policies Service provides a new grpc api which can be used to return whether a requested operation is allowed or not.
> Open Policy Agent is used to determine the set of rules of what is permitted and what is not.
> 
> Policies are written in the [rego query language](https://www.openpolicyagent.org/docs/latest/policy-language/).
> 
> The Policies Service consists of the following modules
> 
> * Proxy Authorization (middleware)
> * Event Authorization (async post-processing)
> * GRPC API (can be used from other services too)
> 

Reviewer 2: ```suggestion
The Policies Service provides a new grpc api which can be used to return whether a requested operation is allowed or not. To do so, Open Policy Agent (OPA) is used to determine the set of rules of what is permitted and what is not.

Policies are written in the [rego query language](https://www.openpolicyagent.org/docs/latest/policy-language/). The location of the rego files can be configured via yaml, a configuration via environment variables is not possible.

The Policies Service consists of the following modules:

*   Proxy Authorization (middleware)
*   Event Authorization (async post-processing)
*   GRPC API (can be used from other services too)

To configure the Policies Service, three environment variables need to be defined:

*   `POLICIES_ENGINE_TIMEOUT`
*   `POLICIES_POSTPROCESSING_QUERY`
*   `PROXY_POLICIES_QUERY`

Note that each query setting defines the [Complete Rules](https://www.openpolicyagent.org/docs/latest/#complete-rules) variable defined in the rego rule set the corresponding step uses for the evaluation. If the variable is mistyped or not found, the evaluation defaults to deny. Individual query definitions can be defined for each module.

To activate a the policies service for a module, it must be started with a yaml configuration that points to one or more rego files. Note that if the service is scaled horizontally, each instance should have access to the same rego files to avoid unpredictable results.
```

> ### GRPC service
> 
> This service can be used from any other internal service, it can also be used for example by third parties to find out if an action is allowed or not.
> This layer is already used by the proxy middleware.
> 
> ### Event service
> 
> This layer is event-based and part of the asynchronous post-processing, since processing at this point is asynchronous, the operations there can also take longer and be more expensive,
> the bytes of a file can be examined here as an example.
> 
> ### Proxy Middleware
> 
> The [ocis proxy](../proxy) already includes such a middleware which uses the [GRPC service](#grpc-service) to evaluate the policies by using a configurable query.
> Since the Proxy is in heavy use and every request is processed here, only simple decisions that can be processed quickly are recommended here, more complex queries such as file evaluation should be avoided urgently.
> 

Reviewer 2: ```suggestion
## Modules

### GRPC Service

This service can be used from any other internal service. It can also be used for example by third parties to find out if an action is allowed or not. This layer is already used by the proxy middleware.

### Event Service

This layer is event-based and part of the postprocessing service. Since processing at this point is asynchronous, the operations can also take longer and be more expensive, like evaluating the bytes of a file.

### Proxy Middleware

The [ocis proxy](../proxy) already includes such a middleware which uses the [GRPC service](#grpc-service) to evaluate the policies by using a configurable query. Since the Proxy is in heavy use and every request is processed here, only simple and quick decisions should be evaluated. More complex queries such as file evaluation are strongly discouraged.

```
I am not clear about the GRPC Service as there is nothing that can be configured. This means, in which cases is the policies service returing results to it? How can the policies service be understood when it comes to GRPC? The intro text says `can be used from other services` but what is meant with that. This needs more clarification.

Author: not sure if we should document it, the GRPC service is really low level, most of our services can be used via GRPC, but to do so you have to read code.

> ## Example
> The Policies Service contains a set of pre-configured example policies, thos policies can be found in the [examples directory](./examples).
> The contained policies disallows ocis to create certain filetypes, both for the proxy middleware and the events service.
> 
> To use the example policies, it's required to configure ocis to use these files which can be done by adding:
> 

Reviewer 2: ```suggestion
## Example Policies

The policies service contains a set of pre-configured example policies. Those policies can be found in the [examples directory](./examples). The contained policies disallows ocis to create certain filetypes, both for the proxy middleware and the events service.

To use the example policies, it's required to configure ocis to use these files which can be done by adding:

```
With regards to the examples directory, this should be moved to https://github.com/owncloud/ocis/tree/master/deployments/examples
(like rego example files). This base directory is already available from both the dev and admin documentation. Else this is only located in the service and hard to find/reference.

Author: not sure if we shoul move those there, those examples currently contain only deployments.
I think the place is ok for me, or we re-locate the old examples and give them a more meaningful naming.

Thoughts? cc. @wkloucek

Reviewer 2: Note to give admins a quickstart, these examples help a lot - if they can be accessed easily (and the link proposed is already working). Naming is imho secondary. We can do some description in the admin docs.

```yaml
policies:
  engine:
    policies:
      - YOUR_PATH/examples/policies/proxy.rego
      - YOUR_PATH/examples/policies/postprocessing.rego
      - YOUR_PATH/examples/policies/utils.rego
```

> Once the policies are configured correctly we need to set up the correct queries for the proxy middleware and for the events service.
> 

Reviewer 2: ```suggestion
Once the policies are configured correctly, the _QUERY configuration needs to be defined for the proxy middleware and for the events service.

```

### Proxy

```yaml
proxy:
  policies_middleware:
    query: data.proxy.granted
```

the same can be achieved by setting the `PROXY_POLICIES_QUERY=data.proxy.granted` environment variable.

### ASYNC postprocessing

```yaml
policies:
  postprocessing:
    query: data.postprocessing.granted
```

> the same can be achieved by setting the `POLICIES_POSTPROCESSING_QUERY=data.postprocessing.granted` environment variable.
> As soon as that query is configured correctly, postprocessing must be informed to use the policies step too by setting the environment variable `POSTPROCESSING_STEPS=policies`.
> 
> 
> 
> 
> 
> 

Reviewer 2: ```suggestion
The same can be achieved by setting the `POLICIES_POSTPROCESSING_QUERY=data.postprocessing.granted` environment variable. As soon as that query is configured correctly, postprocessing must be informed to use the policies step by setting the environment variable `POSTPROCESSING_STEPS=policies`. Note that additional steps can be configured and their appearance defines the order of processing. For details see the postprocessing service documentation.
```
